kms/tpmkms/tpmkms.go

```diff
@@ -534,58 +534,33 @@ func (k *TPMKMS) CreateAttestation(req *apiv1.CreateAttestationRequest) (*apiv1.
 	ekKeyURL := ekURL(ekKeyID)
 	permanentIdentifier := ekKeyURL.String()
 
-	if properties.ak {
-		// TODO(hs): decide if we actually want to support this case? TPM attestation
-		// is about attesting application keys using attestation keys.
-		ak, err := k.tpm.GetAK(ctx, properties.name)
+	// check if the derived EK URI fingerprint representation matches the provided
+	// permanent identifier value. The current implementation requires the EK URI to
+	// be used as the AK identity, so an error is returned if there's no match. This
+	// could be changed in the future, so that another attestation flow takes place,
+	// instead, for example.
+	if k.permanentIdentifier != "" && permanentIdentifier != k.permanentIdentifier {
+		return nil, fmt.Errorf("the provided permanent identifier %q does not match the EK URL %q", k.permanentIdentifier, permanentIdentifier)
+	}
+
+	var key *tpm.Key
+	akName := properties.name
+	if !properties.ak {
+		key, err = k.tpm.GetKey(ctx, properties.name)
 		if err != nil {
 			return nil, err
 		}
-		akPub := ak.Public()
-		if akPub == nil {
-			return nil, fmt.Errorf("failed getting AK public key")
-		}
-		akChain := ak.CertificateChain()
-		if len(akChain) == 0 {
-			return nil, fmt.Errorf("no certificate chain available for AK %q", properties.name)
+		if !key.WasAttested() {
+			return nil, fmt.Errorf("key %q was not attested", key.Name())
 		}
-		// TODO(hs): decide if we want/need to return these; their purpose is slightly
-		// different from the key certification parameters.
-		_, err = ak.AttestationParameters(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed getting AK attestation parameters: %w", err)
-		}
-		return &apiv1.CreateAttestationResponse{
-			Certificate:         akChain[0], // certificate for the AK
-			CertificateChain:    akChain,    // chain for the AK, including the leaf
-			PublicKey:           akPub,      // returns the public key of the attestation key
-			PermanentIdentifier: permanentIdentifier,
-		}, nil
+		akName = key.AttestedBy()
```

Comment:
Looks good, but I think it would be clear if you split the logic in two (if/else). They only share the call to `tpm.GetAK()`

```golang
if !provisioner.ak {
    ...
    // Do we need this? 
    // We already know the key was attested. 
    // And we're not using it in this flow.
    ak, err := k.tpm.GetAK(ctx, akName)
    if err != nil {
        return nil, err
    }
    ...
    return  &apiv1.CreateAttestationResponse{}, nil
}

// AK code
ak, err := k.tpm.GetAK(ctx, properties.name)
if err != nil {
    return  nil, err
}
...
return  &apiv1.CreateAttestationResponse{}, nil
```

Reply:
Missed a code block.